parser.rst

```diff
@@ -0,0 +1,884 @@
+===========================
+
+:Author: Pablo Galindo Salgado
+
+.. highlight:: none
+
+Abstract
+--------
+
+The Parser in CPython is currently a `PEG (Parser Expression Grammar)
+<https://en.wikipedia.org/wiki/Parsing_expression_grammar>`_ parser.  The first
+version of the parser used to be an `LL(1)
+<https://en.wikipedia.org/wiki/LL_parser>`_ based parser that was one of the
+oldest parts of CPython implemented before it was replaced by :pep:`617`. In
+particular, both the current parser and the old LL(1) parser are the output of a
+`parser generator <https://en.wikipedia.org/wiki/Compiler-compiler>`_. This
+means that the way the parser is written is by feeding a description of the
+Grammar of the Python language to a special program (the parser generator) which
+outputs the parser. The way the Python language is changed is therefore by
+modifiying the grammar file and developers rarely need to interact with the
+parser generator itself other than use it to generate the parser.
+
+How PEG Parsers Work
+--------------------
+
+.. _how-peg-parsers-work:
+
+A PEG (Parsing Expression Grammar) grammar differs from a context-free grammar
+(like the current one) in the fact that the way it is written more closely
+reflects how the parser will operate when parsing it. The fundamental technical
+difference is that the choice operator is ordered. This means that when writing::
+
+  rule: A | B | C
+
+a context-free-grammar parser (like an LL(1) parser) will generate constructions
+that given an input string will *deduce* which alternative (``A``, ``B`` or ``C``)
+must be expanded, while a PEG parser will check if the first alternative succeeds
+and only if it fails, will it continue with the second or the third one in the
+order in which they are written. This makes the choice operator not commutative.
+
+Unlike LL(1) parsers, PEG-based parsers cannot be ambiguous: if a string parses,
+it has exactly one valid parse tree. This means that a PEG-based parser cannot
+suffer from the ambiguity problems described in the previous section.
+
+PEG parsers are usually constructed as a recursive descent parser in which every
+rule in the grammar corresponds to a function in the program implementing the
+parser and the parsing expression (the "expansion" or "definition" of the rule)
+represents the "code" in said function. Each parsing function conceptually takes
+an input string as its argument, and yields one of the following results:
+
+* A "success" result. This result indicates that the expression can be parsed by
+  that rule and the function may optionally move forward or consume one or more
+  characters of the input string supplied to it.
+* A "failure" result, in which case no input is consumed.
+
+Notice that "failure" results do not imply that the program is incorrect or a
+parsing failure because as the choice operator is ordered, a "failure" result
+merely indicates "try the following option". A direct implementation of a PEG
```

Comment:
This seems directed at people who are so enmeshed in current parsers (as oppose to programming), that sequential rather thanparallel testing or matching seems weird rather than natural.

Reply:
In the word or grammars the "|" operator doesn't imply parallel or sequential testing because the grammar is *not* describing how strings in the language will be matched. What makes PEG unique is they the parser is describing how to generate a parser, so there is a much closer relationship between the grammar and "execution".